### ISSUE TITLE
minimize function from jax.scipy.optimize does not converge

### ISSUE THREAD

**Issue by sushmit86** [google/jax#19078]
### Description

I have a optimization function which converges using tfp (jax substrate ) however it wont converge when I am using jax scipy optimize.
Here is the code below

```
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import jax
plt.style.use('fivethirtyeight')
import jax.numpy as jnp
from tensorflow_probability.substrates import jax as tfp_jax
import plotly.express as px
from jax.scipy.optimize import minimize
tfd_jax = tfp_jax.distributions

W = 6
L = 3

key = jax.random.PRNGKey(11)
dist = tfd_jax.JointDistributionNamed(
    {
        "probability":tfd_jax.Uniform(low= 0.0, high = 1.0),
        "water": lambda probability:tfd_jax.Binomial(total_count = W +L,probs = probability)
    }
)
def neg_log_prob(x):
    return tfp_jax.math.value_and_gradient(
        lambda p: -dist.log_prob(
            water=W, probability= jnp.clip(p[-1],0,1)),
        x,
    )


optim_results = tfp_jax.optimizer.bfgs_minimize(neg_log_prob,
                                                initial_position=[0.5],
                                                tolerance = 1e-6)
assert(optim_results.converged) # The results converge here
# Now using the minimize function
test_fn = lambda x: -dist.log_prob(water=W, probability= jnp.clip(x,0,1))[0]

optim_scipy_results = minimize(test_fn,x0 = jnp.array([0.5]),method="BFGS",tol=1e-6)
assert(optim_scipy_results.success) # This fails
```

### What jax/jaxlib version are you using?

0.4.23

### Additional system info?

using colab

**Comment by jakevdp** [comment]
I suspect the problem is that your test function is very poorly behaved:
```python
x = jnp.linspace(-5, 5)
prob = jax.vmap(test_fn)(x[:, None])
print(prob)
```
```
[      inf       inf       inf       inf       inf       inf       inf
       inf       inf       inf       inf       inf       inf       inf
       inf       inf       inf       inf       inf       inf       inf
       inf       inf       inf       inf 9.586372  3.7681832 1.7481494
 1.3463035 3.5966973       inf       inf       inf       inf       inf
       inf       inf       inf       inf       inf       inf       inf
       inf       inf       inf       inf       inf       inf       inf
       inf]
```
Numerical optimizers in general have difficulty if there are large regions of the parameter space with no useful gradient.

It seems like you'd need to apply some bounds to your optimization problem; there's no bounded optimization built-in to JAX, but you might have luck with constrained optimizers in the `jaxopt` package: https://jaxopt.github.io/stable/constrained.html

Best of luck!
